### PR TITLE
Support limiting the buffer size in LargeArrayBuilder.

### PR DESCRIPTION
**Pull request by jamesqo**
Continuation of: #14006. Unfortunately, I have a habit of force pushing my commits, so I wasn't able to reopen the old PR. Anyway #12703 was merged this morning so this is no longer blocked; this needs #13628 to apply some optimizations to `Take`, but if this is merged beforehand, I can just update that PR with those changes instead (and vice versa).

---

### Description from original PR

There are certain Linq methods, such as `Where` and `Take`, where we cannot pinpoint the exact number of elements the iterator will contain, but we can establish an upper bound on how many elements there are. For example, `list.Where(p)` can have at most `list.Count` elements, and `e.Take(5)` can have at most 5 elements.

Currently, when we call `ToArray` on any of these iterators, the resizing pattern will be to allocate 4, then 8, then 16, etc. size buffers. This can be wasteful because e.g. if the source enumerable in `Where` contains 100 elements, we allocate space for 128 elements even though those last 28 slots will never be used.

This PR adds support to `LargeArrayBuilder<T>` to limit how many elements we allocate, so there is no extra space wasted. It should help drive down allocations in `Where`, `Where.Select`, & `Take` when the maximum count is far from the next power of 2.

---

[Performance test][1] / [analysis](https://gist.github.com/jamesqo/0ca128f95f59cd434f7ec544ce09377e)

There are major reductions in GCs (about 25%) when the length of `_source` is just above a power of 2, and the predicate returns true for every item.

cc @JonHanna @VSadov @stephentoub 

[1]: https://github.com/jamesqo/Dotnet/blob/fd26359a7b7d53b0598887ab58f54d591f3b7b40/Program.cs

### CODE REVIEW

File: src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs [hunk 1]
```diff
@@ -20,21 +20,36 @@ internal struct LargeArrayBuilder<T>
         private T[] _current;               // Current buffer we're reading into. If _count <= ResizeLimit, this is _first.
         private int _index;                 // Index into the current buffer.
         private int _count;                 // Count of all of the items in this builder.
+        private int _maxCapacity;           // The maximum capacity this builder can have.
```
Comment (stephentoub) [inline src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs]: readonly?

File: src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs [hunk 2]
```diff
@@ -53,6 +68,8 @@ public LargeArrayBuilder(bool initialize)
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public void Add(T item)
         {
+            Debug.Assert((uint)_maxCapacity > (uint)_count);
```
Comment (stephentoub) [inline src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs]: Why the cast to uint for _maxCapacity?

File: src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs [hunk 3]
```diff
@@ -161,7 +182,9 @@ private void AllocateBuffer()
             //   and start reading into _current. Set _index to 0.
             // - When _current runs out of space, add it to _buffers and repeat the
             //   above step, except with _current.Length * 2.
+            // - Make sure we never pass _maxCapacity in all of the above steps.
 
+            Debug.Assert((uint)_maxCapacity > (uint)_count);
```
Comment (stephentoub) [inline src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs]: Same cast question

File: src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs [hunk 4]
```diff
@@ -187,8 +211,22 @@ private void AllocateBuffer()
...
+                    // Note: In case _count overflows (last buffer is 0x40000000) or _count * 2
+                    // overflows (last buffer is 0x20000000), we must cast to uint for correctness
+                    // during comparisons.
+
+                    Debug.Assert((uint)_count >= (uint)ResizeLimit * 2);
+                    Debug.Assert(_count == _current.Length * 2);
+
                     _buffers.Add(_current);
-                    nextCapacity = _current.Length * 2;
+                    nextCapacity = (int)Math.Min((uint)_count * 2, (uint)_maxCapacity) - _count;
```
Comment (stephentoub) [inline src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs]: Why not:
```C#
Math.Min(_count, _maxCapacity - _count)
```
?

File: src/Common/tests/Tests/System/Collections/Generic/LargeArrayBuilderTests.cs [hunk 5]
```diff
@@ -98,6 +112,24 @@ public static TheoryData<IEnumerable<T>> EnumerableData()
...
+
+            var enumerables = EnumerableData().Select(array => array[0]).Cast<IEnumerable<T>>();
+
+            foreach (IEnumerable<T> enumerable in enumerables)
+            {
+                int count = enumerable.Count();
+                data.Add(enumerable, count);
+
+                data.Add(enumerable, count / 2);
+                data.Add(enumerable, count / 4);
```
Comment (stephentoub) [inline src/Common/tests/Tests/System/Collections/Generic/LargeArrayBuilderTests.cs]: Why are these necessary?

File: src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs [hunk 6]
```diff
@@ -187,8 +211,22 @@ private void AllocateBuffer()
...
+                    // As you can see, our count will be just double the last buffer.
+                    // Now, say _maxCapacity is 500. We will find the right amount to allocate by
+                    // doing min(256 * 2, 500) - 256. This represents the total we've allocated after
+                    // this operation, minus the amount we've already allocated.
+
+                    // Note: In case _count overflows (last buffer is 0x40000000) or _count * 2
+                    // overflows (last buffer is 0x20000000), we must cast to uint for correctness
+                    // during comparisons.
+
+                    Debug.Assert((uint)_count >= (uint)ResizeLimit * 2);
```
Comment (stephentoub) [inline src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs]: I don't understand why these casts are necessary.  We know that ResizeLimit is an int... it's a const int defined earlier, and it's defined to be positive.  We also know from the earlier assert that _count < _maxCapacity, and _maxCapacity has already been asserted to be greater than 0.  So if _count is negative, an assert would have already failed.

File: src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs (hunk 3 above)
Comment (jamesqo) [inline src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs]: I think we may need to keep the uint cast here. We could have come here from `AddRange`, where we set `_count` to int.MinValue (last buffer is 0x40000000). If that were the case and we removed these casts, the assert will pass even though we've exceeded our limit. It's different from `Add` where we didn't modify that field yet.

Comment (stephentoub) [inline src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs]: Also, are you saying that there is a legitimate case at run-time where this assert could fail?  If so, that shouldn't be an assert.

File: src/Common/tests/Tests/System/Collections/Generic/LargeArrayBuilderTests.cs [hunk 7]
```diff
@@ -98,6 +112,21 @@ public static TheoryData<IEnumerable<T>> EnumerableData()
             return data;
         }
 
+        public static TheoryData<IEnumerable<T>, int> MaxCapacityData()
+        {
+            var data = new TheoryData<IEnumerable<T>, int>();
+
+            var enumerables = EnumerableData().Select(array => array[0]).Cast<IEnumerable<T>>();
```
Comment (stephentoub) [inline src/Common/tests/Tests/System/Collections/Generic/LargeArrayBuilderTests.cs]: Nit: var => ```IEnumerable<IEnumerable<T>>```

File: src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs (hunk 3 above)
Comment (jamesqo) [inline src/Common/src/System/Collections/Generic/LargeArrayBuilder.cs]: > But why is the cast needed on _maxCapacity?

Force of habit; I always cast both sides to `uint` otherwise Roslyn converts both sides to `long` which can be unintended in some cases. It does not have anything to do with checking if `_maxCapacity >= 0`.

> Also, are you saying that there is a legitimate case at run-time where this assert could fail?

No; the assert will only fail if `AddRange` causes it to add past `_maxCapacity`, which I mentioned is prohibited in the XML docs.

**Comment by stephentoub** [comment]
Thanks. LGTM.
